Docs/Testing.md

```diff
@@ -0,0 +1,287 @@
+}
+```
+
+Unlike XFAIL tests in `lit.py`, tests that are expected to fail will _not_ execute during the build. If your test is disabled in this manner, you should investigate the bug by running the test suite locally; you can do so without changing the source code by setting the `NS_FOUNDATION_ATTEMPT_XFAIL_TESTS` environment variable set to the string `YES`, which will cause tests that were disabled this way to attempt to execute anyway.
+
+## Test Internal Behavior Carefully: `@testable import`
+
+### In brief
+
+* Prefer black box (contract-based) testing to white box testing wherever possible
+* Some contracts cannot be tested or cannot be tested reliably (for instance, per-platform fallback paths); it is appropriate to use `@testable import` to test their component parts instead
+* Ensure your test wraps any reference to `internal` functionality with `#if NS_FOUNDATION_ALLOWS_TESTABLE_IMPORT`
+* Ensure the file you are using this in adds a `@testable import` prologue
+* Run with this enabled _and_ disabled prior to PR
+
+### Why and How
+
+In general, we want to ensure that tests are written to check the _contract_ of the API, [as documented for each class](https://developer.apple.com/). It is of course acceptable to have the test implementation be informed by the implementation, but we want to make sure that tests still make sense if we replace an implementation entirely, [as we sometimes do](https://github.com/apple/swift-corelibs-foundation/pull/2331).
+
+This doesn't always work. Sometimes the contract specifies that a certain _result_ will occur, and that result may be platform-specific or trigger in multiple ways, all of which we'd like to test (for example, different file operation paths for volumes with different capabilities). In this case, we can reach into Foundation's `internal` methods by using `@testable import` and test the component parts or invoke private API ("SPI") to alter the behavior so that all paths are taken.
+
+If you think this is the case, you must be careful. We run tests both against a debug version of Foundation, which supports `@testable import`, and the release library, which does not. Your tests using `internal` code must be correctly marked so that tests don't succeed in one configuration but fail in the other.
+
+To mark those tests:
+
+* Wrap code using internal features with `#if NS_FOUNDATION_ALLOWS_TESTABLE_IMPORT`:
+
+```swift
+func testSomeFeature() {
+#if NS_FOUNDATION_ALLOWS_TESTABLE_IMPORT
+	try Date._someInternalTestModifierMethod {
+		// …
+	}
+#endif
+}
+```
+
+* In the file you're adding the test, if not present, add the appropriate `@testable import` import prologue at the top. It will look like the one below, but **do not** copy and paste this — instead, search the codebase for `@testable import` for the latest version:
+
+```swift
+// It will look something like this:
+
+#if NS_FOUNDATION_ALLOWS_TESTABLE_IMPORT
+    #if canImport(SwiftFoundation) && !DEPLOYMENT_RUNTIME_OBJC
```

Comment:
I dont think the `!DEPLOYMENT_RUNTIME_OBJC` is actually that useful anymore now that `canImport(SwiftFoundation) ` is used. I cant see a situation where `canImport(SwiftFoundation)` returns true but we DONT import it and instead `import Foundation`

Reply:
True. A patch that fixes this should also fix this documentation in the same swoop.